### PR TITLE
[Model] Add support for the multi-modal Llama 3.2 model

### LINKED ISSUE #8812: [New Model]: Llama 3.2
### The model to consider.
- **Huggingface collection:** https://huggingface.co/collections/meta-llama/llama-32-66f448ffc8c32f949b04c8cf

Highlighted model weights:
- **1B Instruct Model:** https://huggingface.co/meta-llama/Llama-3.2-1B-Instruct
- **3B Instruct Model:** https://huggingface.co/meta-llama/Llama-3.2-3B-Instruct
- **11B Instruct Model:** https://huggingface.co/meta-llama/Llama-3.2-11B-Vision-Instruct
- **90B Instruct Model:** https://huggingface.co/meta-llama/Llama-3.2-90B-Vision-Instruct

### The closest model vllm already supports.

https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/llama.py
https://github.com/vllm-project/vllm/blob/main/vllm/model_executor/models/llava_onevision.py

### What's your difficulty of supporting the model you want?

Yes, Llama 3.2 is multimodal with a different architecture than previous multimodal Llama models.

### Before submitting a new issue...

- [X] Make sure you already searched for relevant issues, and asked the chatbot living at the bottom right corner of the [documentation page](https://docs.vllm.ai/en/latest/), which can answer lots of frequently asked questions.

### PR DESCRIPTION
**Pull request by heheda12345**
This PR adds support for the [Llama3.2](https://ai.meta.com/blog/llama-3-2-connect-2024-vision-edge-mobile-devices)'s vision models. 

~Currently, we are blocked by a HuggingFace Transformers release after https://github.com/huggingface/transformers/pull/33703~

We will make a release as soon as this PR is merged. 

Closes #8812

### CODE REVIEW

File: vllm/model_executor/models/mllama.py [hunk 1]
```diff
@@ -0,0 +1,1132 @@
...
+        num_tiles_height = canvas_height // tile_size
+        num_tiles_width = canvas_width // tile_size
+        num_tiles += num_tiles_height * num_tiles_width
+
+    # set encoder prompt based on num_tiles
+    assert hf_config.vision_config.image_size % 14 == 0, \
+        "chunk size should be multiple of 14"
+    token_per_chunk = (hf_config.vision_config.image_size // 14)**2 + 1
+    num_tokens = num_tiles * token_per_chunk
+    llm_inputs["encoder_prompt"] = "<|image|>" * num_tokens
```
Comment (ywang96) [inline vllm/model_executor/models/mllama.py:116]: Let's make `MLLAMA_IMAGE_TOKEN = "<|image|>"` and move it to the top

File: vllm/model_executor/models/mllama.py [hunk 2]
```diff
@@ -0,0 +1,1132 @@
...
+        tile_position_embedding = tile_position_embedding.reshape(
+            batch_size, self.max_num_tiles, self.num_patches, self.hidden_size)
+        gated_tile_position_embedding = self.gate.tanh(
+        ) * tile_position_embedding
+        hidden_state = hidden_state + gated_tile_position_embedding
+
+        return hidden_state
+
+
+class MllamaVisionSdpaAttention(nn.Module):
```
Comment (ywang96) [inline vllm/model_executor/models/mllama.py:301]: Let's add a `TODO` here to add Vision Attention with other backends since currently we use `xformers` for other VLMs

File: vllm/model_executor/models/mllama.py [hunk 3]
```diff
@@ -0,0 +1,1132 @@
...
+        attn_output = attn_output.transpose(1, 2).contiguous()
+        attn_output = attn_output.reshape(attn_output.shape[0],
+                                          attn_output.shape[1], -1)
+        output, _ = self.o_proj(attn_output)
+        return output
+
+
+class MllamaVisionEncoderLayer(nn.Module):
+
+    def __init__(self, config, is_gated: bool = False):
```
Comment (ywang96) [inline vllm/model_executor/models/mllama.py:357]: Add type hint for config

File: vllm/model_executor/models/mllama.py [hunk 4]
```diff
@@ -0,0 +1,1132 @@
...
+
+
+class MllamaVisionEncoder(nn.Module):
+    """
+    Transformer encoder consisting of `config.num_hidden_layers` self attention 
+    layers. Each layer is a [`MllamaEncoderLayer`].
+
+    Args:
+        config: MllamaConfig
+    """
```
Comment (ywang96) [inline vllm/model_executor/models/mllama.py:408]: Can you update the docstring here? It's not accurate currently

File: vllm/model_executor/models/mllama.py [hunk 5]
```diff
@@ -0,0 +1,1132 @@
...
+        ) * hidden_states
+        return hidden_states
+
+
+class MllamaTextModel(nn.Module):
+    config_class = config_mllama.MllamaTextConfig
+    base_model_prefix = "model"
+    _no_split_modules = [
+        "MllamaCrossAttentionDecoderLayer", "MllamaSelfAttentionDecoderLayer"
+    ]
```
Comment (ywang96) [inline vllm/model_executor/models/mllama.py:773]: Remove these

File: vllm/model_executor/models/mllama.py [hunk 6]
```diff
@@ -0,0 +1,1132 @@
...
+        super().__init__()
+        self.vocab_size = config.text_config.vocab_size
+        self.hidden_size = config.text_config.hidden_size
+        self.max_num_tiles = config.vision_config.max_num_tiles
+        self.vision_output_dim = config.vision_config.vision_output_dim
+        self.pad_token_id = \
+            config.pad_token_id if config.pad_token_id is not None else -1
+        self.image_size = config.vision_config.image_size
+
+        self.vision_model = MllamaVisionModel(config.vision_config, )
```
Comment (ywang96) [inline vllm/model_executor/models/mllama.py:910]:
```suggestion
        self.vision_model = MllamaVisionModel(config.vision_config)
```

File: examples/openai_vision_api_client.py [hunk 7]
```diff
@@ -38,7 +38,7 @@
         "content": [
             {
                 "type": "text",
-                "text": "What’s in this image?"
+                "text": "What's in this image?"
```
Comment (simon-mo) [inline examples/openai_vision_api_client.py:41]:
```suggestion
                "text": "What's in this image?"
```

File: vllm/model_executor/models/mllama.py [hunk 8]
```diff
@@ -0,0 +1,1132 @@
...
+                dtype=cross_attention_states.dtype)
+            start_pos = 0
+            for seq_len, vision_token_in_batch in zip(
+                    attn_metadata.encoder_seq_lens, cross_attention_states):
+                end_pos = start_pos + seq_len
+                cross_attention_states_flat[
+                    start_pos:end_pos] = vision_token_in_batch[:seq_len]
+                start_pos = end_pos
+            cross_attention_states = cross_attention_states_flat
+            cross_attention_mask = None  # TODO
```
Comment (ywang96) [inline vllm/model_executor/models/mllama.py:1074]: Can you add more information for TODO here?

File: examples/openai_vision_api_client.py (hunk 7 above)
Comment (simon-mo) [reply examples/openai_vision_api_client.py:41]: Changing unicode from asian `'` to english `'` to avoid some encoding error.

File: examples/openai_vision_api_client.py [hunk 9]
```diff
@@ -75,7 +75,7 @@ def encode_image_base64_from_url(image_url: str) -> str:
...
```
Comment (simon-mo) [inline examples/openai_vision_api_client.py:78]:
```suggestion
                "text": "What's in this image?"
```

File: tests/models/encoder_decoder/vision_language/test_mllama.py [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+})
+
+text_only_prompts = [
+    "The color of the sky is blue but sometimes it can also be",
+]
+
+models = [
+    "meta-llama/Llama-3.2-11B-Vision-Instruct",
+]
+# TODO: Update model path to huggingface model
```
Comment (simon-mo) [inline tests/models/encoder_decoder/vision_language/test_mllama.py:31]:
```suggestion
```

File: vllm/model_executor/models/mllama.py [hunk 11]
```diff
@@ -0,0 +1,1132 @@
...
+        image_inputs = self._parse_and_validate_image_input(**kwargs)
+        if image_inputs is None:
+            cross_attention_mask = None
+            full_text_row_masked_out_mask = (
+                attn_metadata.encoder_seq_lens_tensor != 0).reshape(-1, 1).to(
+                    input_ids.device)
+            cross_attention_states = None
+            skip_cross_attention = max(attn_metadata.encoder_seq_lens) == 0
+        else:
+            # llama's reference implementation runs the vision model on CPU
```
Comment (ywang96) [inline vllm/model_executor/models/mllama.py:1047]:
```suggestion
            # NOTE: llama's reference implementation runs the vision model
            # on CPU; currently the vision model runs on GPU in vLLM 
```

File: vllm/worker/worker.py [hunk 12]
```diff
@@ -222,10 +222,11 @@ def determine_num_available_blocks(self) -> Tuple[int, int]:
         # of the model.
         self.model_runner.profile_run()
 
-        # Calculate the number of blocks that can be allocated with the
-        # profiled peak memory.
+        # # Calculate the number of blocks that can be allocated with the
+        # # profiled peak memory.
         torch.cuda.synchronize()
         free_gpu_memory, total_gpu_memory = torch.cuda.mem_get_info()
+
```
Comment (simon-mo) [inline vllm/worker/worker.py:229]:
```suggestion
```

File: vllm/worker/worker.py [hunk 13]
```diff
@@ -222,10 +222,11 @@ def determine_num_available_blocks(self) -> Tuple[int, int]:
         # of the model.
         self.model_runner.profile_run()
 
-        # Calculate the number of blocks that can be allocated with the
-        # profiled peak memory.
+        # # Calculate the number of blocks that can be allocated with the
+        # # profiled peak memory.
```
Comment (simon-mo) [inline vllm/worker/worker.py:226]:
```suggestion
        # Calculate the number of blocks that can be allocated with the
        # profiled peak memory.
```

File: vllm/worker/enc_dec_model_runner.py [hunk 14]
```diff
@@ -288,33 +294,47 @@ def profile_run(self) -> None:
         max_mm_tokens = self.mm_registry.get_max_multimodal_tokens(
             self.model_config)
         if max_mm_tokens > 0:
-            raise NotImplementedError(
-                "Multi-modal encoder-decoder models are not supported yet")
+            logger.warning("profile run for multi-modal models")
```
Comment (simon-mo) [inline vllm/worker/enc_dec_model_runner.py:297]:
```suggestion
            logger.info("Starting profile run for multi-modal models.")
```

File: vllm/model_executor/models/mllama.py [hunk 15]
```diff
@@ -0,0 +1,1132 @@
...
+                    attn_metadata.encoder_seq_lens, cross_attention_states):
+                end_pos = start_pos + seq_len
+                cross_attention_states_flat[
+                    start_pos:end_pos] = vision_token_in_batch[:seq_len]
+                start_pos = end_pos
+            cross_attention_states = cross_attention_states_flat
+            cross_attention_mask = None  # TODO
+
+            full_text_row_masked_out_mask = torch.ones(
+                (attn_metadata.num_prefill_tokens, 1), dtype=torch.bool)
```
Comment (ywang96) [inline vllm/model_executor/models/mllama.py:1077]: Nit: can you define a separate function for these? Ideally we should keep the forward pass of the main model clean.

File: vllm/model_executor/models/mllama.py [hunk 16]
```diff
@@ -0,0 +1,1132 @@
...
+            for seq_len, encoder_seq_len in zip(
+                    attn_metadata.seq_lens_tensor.cpu(),
+                    attn_metadata.encoder_seq_lens):
+                if encoder_seq_len == 0:
+                    full_text_row_masked_out_mask[start_pos:start_pos +
+                                                  seq_len] = False
+                start_pos += seq_len
+            full_text_row_masked_out_mask = full_text_row_masked_out_mask.to(
+                cross_attention_states.device)
+            skip_cross_attention = False
```
Comment (ywang96) [inline vllm/model_executor/models/mllama.py:1088]: Ditto

File: vllm/worker/worker.py (hunk 13 above)
Comment (ywang96) [inline vllm/worker/worker.py:226]: Revert this change

File: vllm/worker/worker.py (hunk 12 above)
Comment (ywang96) [inline vllm/worker/worker.py:229]:
```suggestion
```

File: vllm/entrypoints/chat_utils.py [hunk 17]
```diff
@@ -368,6 +371,37 @@ def _parse_chat_message_content_parts(
...
+            if part_type == "text":
+                text = _TextParser(part)["text"]
+                texts.append(text)
+            elif part_type == "image_url":
+                image_url = _ImageParser(part)["image_url"]
+
+                if image_url.get("detail", "auto") != "auto":
+                    logger.warning(
+                        "'image_url.detail' is currently not supported and "
+                        "will be ignored.")
```
Comment (ywang96) [inline vllm/entrypoints/chat_utils.py:391]: IMO we should define a separate function for line 406 - 428 here, and reuse this function to deal with the special case when we need to keep multimodal content.

**Review by ywang96 — COMMENTED** [review]
Thanks for working on this! I left a few comments and please take a look!

File: vllm/transformers_utils/configs/__init__.py [hunk 18]
```diff
@@ -26,11 +27,13 @@
...
     "ExaoneConfig",
+    "MllamaConfig",
     "MLPSpeculatorConfig",
     "NemotronConfig",
     "SolarConfig",
     "UltravoxConfig",
     # Granite can be removed from here once we have upgraded to
     # transformers 4.45+
     "GraniteConfig",
+    ""
```
Comment (ywang96) [inline vllm/transformers_utils/configs/__init__.py:38]:
```suggestion
```

**Comment by mgoin** [comment]
Just leaving this log here as I experiment with quantization on-top of this work:
```
vllm serve mgoin/Llama-3.2-11B-Vision-Instruct-FP8-Dynamic

WARNING 09-25 19:05:44 cuda.py:22] You are using a deprecated `pynvml` package. Please install `nvidia-ml-py` instead, and make sure to uninstall `pynvml`. When both of them are installed, `pynvml` will take precedence and cause errors. See https://pypi.org/project/pynvml for more information.
/home/mgoin/code/vllm/vllm/connections.py:8: RuntimeWarning: Failed to read commit hash:
No module named 'vllm._version'
  from vllm.version import __version__ as VLLM_VERSION
INFO 09-25 19:05:46 api_server.py:526] vLLM API server version dev
INFO 09-25 19:05:46 api_server.py:527] args: Namespace(model_tag='Llama-3.2-11B-Vision-Instruct-FP8-Dynamic', config='', host=None, port=8000, uvicorn_log_level='info', allow_credentials=False, allowed_origins=['*'], allowed_methods=['*'], allowed_headers=['*'], api_key=None, lora_modules=None, prompt_adapters=None, chat_template=None, response_role='assistant', ssl_keyfile=None, ssl_certfile=None, ssl_ca_certs=None, ssl_cert_reqs=0, root_path=None, middleware=[], return_tokens_as_token_ids=False, disable_frontend_multiprocessing=False, enable_auto_tool_choice=False, tool_call_parser=None, model='Llama-3.2-11B-Vision-Instruct-FP8-Dynamic', tokenizer=None, skip_tokenizer_init=False, revision=None, code_revision=None, tokenizer_revision=None, tokenizer_mode='auto', trust_remote_code=False, download_dir=None, load_format='auto', config_format='auto', dtype='auto', kv_cache_dtype='auto', quantization_param_path=None, max_model_len=None, guided_decoding_backend='outlines', distributed_executor_backend=None, worker_use_ray=False, pipeline_parallel_size=1, tensor_parallel_size=1, max_parallel_loading_workers=None, ray_workers_use_nsight=False, block_size=16, enable_prefix_caching=False, disable_sliding_window=False, use_v2_block_manager=False, num_lookahead_slots=0, seed=0, swap_space=4, cpu_offload_gb=0, gpu_memory_utilization=0.9, num_gpu_blocks_override=None, max_num_batched_tokens=None, max_num_seqs=256, max_logprobs=20, disable_log_stats=False, quantization=None, rope_scaling=None, rope_theta=None, enforce_eager=False, max_context_len_to_capture=None, max_seq_len_to_capture=8192, disable_custom_all_reduce=False, tokenizer_pool_size=0, tokenizer_pool_type='ray', tokenizer_pool_extra_config=None, limit_mm_per_prompt=None, mm_processor_kwargs=None, enable_lora=False, max_loras=1, max_lora_rank=16, lora_extra_vocab_size=256, lora_dtype='auto', long_lora_scaling_factors=None, max_cpu_loras=None, fully_sharded_loras=False, enable_prompt_adapter=False, max_prompt_adapters=1, max_prompt_adapter_token=0, device='auto', num_scheduler_steps=1, multi_step_stream_outputs=False, scheduler_delay_factor=0.0, enable_chunked_prefill=None, speculative_model=None, speculative_model_quantization=None, num_speculative_tokens=None, speculative_draft_tensor_parallel_size=None, speculative_max_model_len=None, speculative_disable_by_batch_size=None, ngram_prompt_lookup_max=None, ngram_prompt_lookup_min=None, spec_decoding_acceptance_method='rejection_sampler', typical_acceptance_sampler_posterior_threshold=None, typical_acceptance_sampler_posterior_alpha=None, disable_logprobs_during_spec_decoding=None, model_loader_extra_config=None, ignore_patterns=[], preemption_mode=None, served_model_name=None, qlora_adapter_name_or_path=None, otlp_traces_endpoint=None, collect_detailed_traces=None, disable_async_output_proc=False, override_neuron_config=None, disable_log_requests=False, max_log_len=None, disable_fastapi_docs=False, dispatch_function=<function serve at 0x7c76acb9eb00>)
INFO 09-25 19:05:46 api_server.py:164] Multiprocessing frontend to use ipc:///tmp/0ee642b0-ab8e-4620-b5e2-0dd7f4ba90d5 for IPC Path.
INFO 09-25 19:05:46 api_server.py:177] Started engine process with PID 766178
WARNING 09-25 19:05:46 arg_utils.py:940] The model has a long context length (131072). This may cause OOM errors during the initial memory profiling phase, or result in low performance due to small KV cache space. Consider setting --max-model-len to a smaller value.
WARNING 09-25 19:05:47 cuda.py:22] You are using a deprecated `pynvml` package. Please install `nvidia-ml-py` instead, and make sure to uninstall `pynvml`. When both of them are installed, `pynvml` will take precedence and cause errors. See https://pypi.org/project/pynvml for more information.
/home/mgoin/code/vllm/vllm/connections.py:8: RuntimeWarning: Failed to read commit hash:
No module named 'vllm._version'
  from vllm.version import __version__ as VLLM_VERSION
WARNING 09-25 19:05:49 arg_utils.py:940] The model has a long context length (131072). This may cause OOM errors during the initial memory profiling phase, or result in low performance due to small KV cache space. Consider setting --max-model-len to a smaller value.
INFO 09-25 19:05:49 llm_engine.py:226] Initializing an LLM engine (vdev) with config: model='Llama-3.2-11B-Vision-Instruct-FP8-Dynamic', speculative_config=None, tokenizer='Llama-3.2-11B-Vision-Instruct-FP8-Dynamic', skip_tokenizer_init=False, tokenizer_mode=auto, revision=None, override_neuron_config=None, rope_scaling=None, rope_theta=None, tokenizer_revision=None, trust_remote_code=False, dtype=torch.bfloat16, max_seq_len=131072, download_dir=None, load_format=LoadFormat.AUTO, tensor_parallel_size=1, pipeline_parallel_size=1, disable_custom_all_reduce=False, quantization=compressed-tensors, enforce_eager=False, kv_cache_dtype=auto, quantization_param_path=None, device_config=cuda, decoding_config=DecodingConfig(guided_decoding_backend='outlines'), observability_config=ObservabilityConfig(otlp_traces_endpoint=None, collect_model_forward_time=False, collect_model_execute_time=False), seed=0, served_model_name=Llama-3.2-11B-Vision-Instruct-FP8-Dynamic, use_v2_block_manager=False, num_scheduler_steps=1, multi_step_stream_outputs=False, enable_prefix_caching=False, use_async_output_proc=True, use_cached_outputs=True, mm_processor_kwargs=None)
INFO 09-25 19:05:49 enc_dec_model_runner.py:140] EncoderDecoderModelRunner requires XFormers backend; overriding backend auto-selection and forcing XFormers.
INFO 09-25 19:05:49 selector.py:116] Using XFormers backend.
/home/mgoin/venvs/vllm/lib/python3.10/site-packages/xformers/ops/fmha/flash.py:211: FutureWarning: `torch.library.impl_abstract` was renamed to `torch.library.register_fake`. Please use that instead; we will remove `torch.library.impl_abstract` in a future version of PyTorch.
  @torch.library.impl_abstract("xformers_flash::flash_fwd")
/home/mgoin/venvs/vllm/lib/python3.10/site-packages/xformers/ops/fmha/flash.py:344: FutureWarning: `torch.library.impl_abstract` was renamed to `torch.library.register_fake`. Please use that instead; we will remove `torch.library.impl_abstract` in a future version of PyTorch.
  @torch.library.impl_abstract("xformers_flash::flash_bwd")
INFO 09-25 19:05:51 model_runner.py:1014] Starting to load model Llama-3.2-11B-Vision-Instruct-FP8-Dynamic...
INFO 09-25 19:05:51 selector.py:116] Using XFormers backend.
Loading safetensors checkpoint shards:   0% Completed | 0/3 [00:00<?, ?it/s]
Process SpawnProcess-1:
Traceback (most recent call last):
  File "/usr/lib/python3.10/multiprocessing/process.py", line 314, in _bootstrap
    self.run()
  File "/usr/lib/python3.10/multiprocessing/process.py", line 108, in run
    self._target(*self._args, **self._kwargs)
  File "/home/mgoin/code/vllm/vllm/engine/multiprocessing/engine.py", line 388, in run_mp_engine
    engine = MQLLMEngine.from_engine_args(engine_args=engine_args,
  File "/home/mgoin/code/vllm/vllm/engine/multiprocessing/engine.py", line 138, in from_engine_args
    return cls(
  File "/home/mgoin/code/vllm/vllm/engine/multiprocessing/engine.py", line 78, in __init__
    self.engine = LLMEngine(*args,
  File "/home/mgoin/code/vllm/vllm/engine/llm_engine.py", line 325, in __init__
    self.model_executor = executor_class(
  File "/home/mgoin/code/vllm/vllm/executor/executor_base.py", line 47, in __init__
    self._init_executor()
  File "/home/mgoin/code/vllm/vllm/executor/gpu_executor.py", line 40, in _init_executor
    self.driver_worker.load_model()
  File "/home/mgoin/code/vllm/vllm/worker/worker.py", line 183, in load_model
    self.model_runner.load_model()
  File "/home/mgoin/code/vllm/vllm/worker/model_runner.py", line 1016, in load_model
    self.model = get_model(model_config=self.model_config,
  File "/home/mgoin/code/vllm/vllm/model_executor/model_loader/__init__.py", line 19, in get_model
    return loader.load_model(model_config=model_config,
  File "/home/mgoin/code/vllm/vllm/model_executor/model_loader/loader.py", line 403, in load_model
    model.load_weights(self._get_all_weights(model_config, model))
  File "/home/mgoin/code/vllm/vllm/model_executor/models/mllama.py", line 1123, in load_weights
    param = params_dict[name]
KeyError: 'language_model.model.layers.13.cross_attn.qkv_proj.weight_scale'
```

**Review by ywang96 — APPROVED** [review]
Overall LGTM! Thank you for the contribution!

**Comment by ywang96** [comment]
Latest transformers release broke a few tests on our CI. Investigating...

**Comment by ArthurZucker** [comment]
Anything we can do to help with the failing tests

**Comment by ywang96** [comment]
@ArthurZucker Hey thanks for reaching out and offering the help! 

We already made a release and have a good sense how to fix the failing tests on CI, so hopefully they'll be all resolved during day time here (most of us are in PST), thanks!

**Comment by heheda12345** [comment]
Yes. It is ready now.
